### PR TITLE
Add log events for removed reactions

### LINKED ISSUE #19: [Feature]: Reaction log
### Checked Existing

- [X] I have checked the repository for duplicate issues.

### What feature do you want to see added?

Ability for moderators to see what reactions have been added to a message even if that reaction has been removed

### Why do you want to have this feature?

Some members use reactions to troll or otherwise antagonize people. However, if a user does this and then removes the reaction it can be difficult to tell who in the chat is doing this

### Any other details to share? (OPTIONAL)

_No response_

### PR DESCRIPTION
**Pull request by wolfendale**
Resolves #19 

#### Reaction removal threads
When a reaction is removed from a message, we find the latest event log related to the message and add a message to the "Reaction removal thread" for that message. 

As we currently only add events to the log when messages are updated, if there is no existing event for a message a "Reactions removed" log will be added and used as the basis for the initial thread.

Doing this means that:
- We have a log of what the message content was at the time the reaction was removed
- Each reaction removal is timestamped
- As each "Reaction removal thread" is archived upon creation and after each new message, it never appears in the channel list and never contributes towards our server thread limit

##### "Reactions removed" log example with the "Reaction removal thread" shown
image

##### "Reaction removal thread"
This shows the three different kinds of reactions:
1. Regular emoji reaction
2. Custom emoji reaction from the server that Chubby is in
3. Custom emoji reaction from a different server, these are displayed differently as Chubby cannot reproduce the emoji directly without being a member of the server the emoji is from
image

#### Message events context menu
Especially now we're adding extra things which can cause an event log for a particular message, it can be tough finding the relevant event log for a message. To assist with this, I've added a context menu command for messages which will display links to all of the log events for a particular message.

image

- [x] I have read and agreed to the [Code of Conduct](https://github.com/PretendoNetwork/Pretendo/blob/master/.github/CODE_OF_CONDUCT.md).
- [x] I have read and complied with the [contributing guidelines](https://github.com/PretendoNetwork/Pretendo/blob/master/.github/CONTRIBUTING.md).
- [x] What I'm implementing was an [approved issue](../issues?q=is%3Aopen+is%3Aissue+label%3Aapproved).
- [x] I have tested all of my changes.

### CODE REVIEW

**Review by MatthewL246 — COMMENTED** [review]
Looks good, except for one potential crash. When lots of reactions are removed at once on a message that doesn't have an audit log entry yet, it will crash with the following error.

```js
DiscordAPIError[160004]: A thread has already been created for this message
    at handleErrors (/root/source/chubby/node_modules/@discordjs/rest/dist/index.js:730:13)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async SequentialHandler.runRequest (/root/source/chubby/node_modules/@discordjs/rest/dist/index.js:1133:23)
    at async SequentialHandler.queueRequest (/root/source/chubby/node_modules/@discordjs/rest/dist/index.js:963:14)
    at async _REST.request (/root/source/chubby/node_modules/@discordjs/rest/dist/index.js:1278:22)
    at async GuildTextThreadManager.create (/root/source/chubby/node_modules/discord.js/src/managers/GuildTextThreadManager.js:76:18)
    at async Client.reactionRemoveHandler (/root/source/chubby/dist/events/reactionRemove.js:60:47)
Emitted 'error' event on Client instance at:
    at emitUnhandledRejectionOrErr (node:events:402:10)
    at process.processTicksAndRejections (node:internal/process/task_queues:84:21) {
  requestBody: {
    files: undefined,
    json: {
      name: 'Reaction removal thread',
      auto_archive_duration: undefined,
      type: 11,
      invitable: undefined,
      rate_limit_per_user: undefined
    }
  },
  rawError: {
    message: 'A thread has already been created for this message',
    code: 160004
  },
  code: 160004,
  status: 400,
  method: 'POST',
  url: 'https://discord.com/api/v10/channels/1256770579302973595/messages/1266857725850292285/threads'
}
```

File: src/handlers/context-menu-handler.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+	const contextMenus = interaction.client.contextMenus;
+	const contextMenu = contextMenus.get(commandName);
+
+	// do nothing if no context menu
+	if (!contextMenu) {
+		throw new Error(`Missing command handler for \`${commandName}\``);
+	}
+
+	// run the context menu
```
Comment (jonbarrow) [inline src/handlers/context-menu-handler.ts:14]:
```suggestion
	// * Do nothing if no context menu
	if (!contextMenu) {
		throw new Error(`Missing command handler for \`${commandName}\``);
	}

	// * Run the context menu
```

Comment (wolfendale) [reply src/handlers/context-menu-handler.ts]: I've updated this and also fixed it in the command handler

File: src/events/reactionRemove.ts [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+		});
+		await thread.setArchived(true);
+	}
+}
+
+async function createThread(audit: Message): Promise<ThreadChannel> {
+	try {
+		return await audit.startThread({ name: 'Event detail thread', reason: 'Thread for adding reaction removal events' });
+	} catch {
+		return audit.thread!;
```
Comment (MatthewL246) [inline src/events/reactionRemove.ts:78]: This could use a comment explaining why this is necessary.

**Review by MatthewL246 — APPROVED** [review]
Everything looks good to me.

**Review by jonbarrow — APPROVED** [review]
LGTM

**Review by MatthewL246 — APPROVED** [review]

**Pull request merged**